src/main/java/com/yelp/nrtsearch/server/luceneserver/warming/Warmer.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2020 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.warming;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import com.yelp.nrtsearch.server.grpc.SearchRequest;
+import com.yelp.nrtsearch.server.luceneserver.IndexState;
+import com.yelp.nrtsearch.server.luceneserver.SearchHandler;
+import com.yelp.nrtsearch.server.utils.Archiver;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.lucene.util.NamedThreadFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Warmer {
+  private static final Logger logger = LoggerFactory.getLogger(Warmer.class);
+  private static final String WARMING_QUERIES_RESOURCE = "_warming_queries";
+  public static final String WARMING_QUERIES_DIR = "warming_queries";
+  private static final String WARMING_QUERIES_FILE = "warming_queries.txt";
+
+  private final Archiver archiver;
+  private final String service;
+  private final String resource;
+  private final List<SearchRequest> warmingRequests;
+  private final ReservoirSampler reservoirSampler;
+
+  public Warmer(Archiver archiver, String service, String index, int maxWarmingQueries) {
+    this.archiver = archiver;
+    this.service = service;
+    this.resource = index + WARMING_QUERIES_RESOURCE;
+    this.warmingRequests = Collections.synchronizedList(new ArrayList<>(maxWarmingQueries));
+    this.reservoirSampler = new ReservoirSampler(maxWarmingQueries);
+  }
+
+  public void addSearchRequest(SearchRequest searchRequest) {
+    ReservoirSampler.SampleResult sampleResult =
+        reservoirSampler.sample(searchRequest.getIndexName());
```

Comment:
Since a warmer is already bound to an index, does the sampler need to support multiple indices?